File: python/searching/exponentialsearch.py
'''
   Exponenetial search works by finding the range in which the desired element is present and performing binary search in that given range
'''

def binarySearch( arr, start, end, x): #Recusrively call binary range in the given range
    if end >= start: 
        mid = int(start + ( end-start ) / 2)
        if arr[mid] == x: 
            return mid 
           
        if arr[mid] > x: 
            return binarySearch(arr,start,mid - 1, x) 

        else: 
            return binarySearch(arr, mid + 1, end, x) 
           
    return -1 #If element is not present in the array
  
def exponentialSearch(arr, n, x): #Function to perform exponential search 
    if arr[0] == x:#If element is present at the first pos 
        return 0
          
    i = 1
    while i < n and arr[i] <= x: #Keep doubling to find the range in which the number is present 
        i = i * 2
      
    return binarySearch( arr, i / 2,min(i, n), x) 
      
  
def main():#driver function
    arr= []
    n = int(input("Enter the number of elements: "))
    for i in range(n): #Taking input from the user
        ele = int(input("Enter the element: "))
        arr.append(ele)
    x = int(input("Enter the element to be searched: "))
    result = exponentialSearch(arr, n, x) 
    if result == -1: 
        print("Element not found in the array")
    else: 
        print("Element present at index: ",result) 

main() 

File: cpp/searching/Exponential_Search.cpp
#include<bits/stdc++.h>
using namespace std;

int binarySearch(int arr[],int l,int r,int x){
  if (r >= l) {
    int mid = l + (r - l) / 2;
    if (arr[mid] == x)
      return mid;
    if (arr[mid] > x)
      return binarySearch(arr, l, mid - 1, x);
    return binarySearch(arr, mid + 1, r, x);
  }
  return -1;
}

int exponentialSearch(int arr[], int n, int x) {
  if (arr[0] == x)
    return 0;
  int i = 1;
  while (i < n && arr[i] <= x)
    i = i * 2;
  return binarySearch(arr, i / 2, min(i, n), x);
}

int main() {
  int n;
  cout << "Enter the number of elements: ";
  cin >> n;
  int arr[n];
  cout << "Enter the elements value:" << endl;
  for (int i = 0; i < n; i++) {
    cin >> arr[i];
  }
  int x;
  cout << "Enter the element to be searched: ";
  cin >> x;
  int index = exponentialSearch(arr, n, x);
  if (index != -1)
    cout << "Element found at index " << index;
  else
    cout << "Element not found.";
  return 0;
}